packages/database/src/migrations/20211128090124-MergeFlutrackingResponses-modifies-data.js

```diff
@@ -0,0 +1,189 @@
+  const { rows: answers } = await db.runSql(`
+    select * from answer where survey_response_id IN (${arrayToDbString(responseIds)});
+  `);
+  return groupBy(answers, 'survey_response_id');
+};
+
+const processResponseGroups = async (db, responseGroupBatch) => {
+  const a = Date.now();
+
+  const answerByResponseId = await selectAnswersForResponseGroups(db, responseGroupBatch);
+  const mergeData = responseGroupBatch.map(responseGroup =>
+    mergeAnswersInResponseGroup(responseGroup, answerByResponseId),
+  );
+
+  const responseIdsToDelete = mergeData.map(r => r.responseIdsToDelete).flat();
+  const answersToCreate = mergeData.map(d => d.answersToCreate).flat();
+  await deleteResponses(db, responseIdsToDelete);
+  await createAnswers(db, answersToCreate);
+};
+
+const groupResponsesByEntityAndWeek = duplicateResponses => {
+  const groupedResponses = {};
+
+  duplicateResponses.forEach(record => {
+    const {
+      id1,
+      id2,
+      end_time1: endTime1,
+      end_time2: endTime2,
+      entity_id: entityId,
+      iso_week: isoWeek,
+    } = record;
+    const key = `${entityId}___${isoWeek}`;
+
+    if (!(key in groupedResponses)) {
+      groupedResponses[key] = {};
+    }
+    groupedResponses[key][id1] = { id: id1, endTime: endTime1 };
+    groupedResponses[key][id2] = { id: id2, endTime: endTime2 };
+  });
+
+  return Object.values(groupedResponses).map(group => Object.values(group));
+};
+
+const selectDuplicateFlutrackingResponses = async (db, surveyCode) => {
+  const { rows } = await db.runSql(`
+    select
+      sr.id as id1,
+      sr2.id as id2,
+      sr.end_time as end_time1,
+      sr2.end_time as end_time2,
+      sr.entity_id,
+      concat(date_part('year', sr.data_time), 'W', date_part('week', sr.data_time)) as iso_week
+    from survey_response sr
+    join survey_response sr2 on sr2.id > sr.id and sr2.survey_id = sr.survey_id and sr2.entity_id = sr.entity_id
+    join survey s on s.id = sr.survey_id
+    where
+      s.code = '${surveyCode}' and
+      -- Join on same ISO week
+      date_part('year', sr.data_time) = date_part('year', sr2.data_time) and
+      date_part('week', sr.data_time) = date_part('week', sr2.data_time);
+  `);
+
+  return rows;
+};
+
+const mergeFlutrackingResponses = async (db, surveyCode) => {
+  console.log(`Merging duplicate responses for ${surveyCode}`);
+
+  const duplicateResponses = await selectDuplicateFlutrackingResponses(db, surveyCode);
+  const responseGroups = groupResponsesByEntityAndWeek(duplicateResponses);
+
+  console.log(
+    'Processing response groups in batches. Each batch will be processed as a transaction',
+  );
+  console.log(`Batches found: ${responseGroups.length}`);
+  await processInBatches(responseGroups, RESPONSE_GROUP_BATCH_SIZE, async (responseGroup, i) => {
+    await db.runSql('START TRANSACTION');
+    printProgress(i, responseGroups.length);
+    await processResponseGroups(db, responseGroup);
+    await db.runSql('COMMIT');
```

Comment:
This is kind of required for two reasons:
1. If we don't commit the current changes, memory builds up and quickly brings the system to its knees
2. This is a long running migration so it's probably desired to commit changes for each batch (data for each of them is independent). This way, even if the migration stop halfway through, we can run `yarn  migrate` again and it will continue from the last "checkpoint".

Reply:
Yes great pattern! I understood the migration as a whole to be wrapped in a transaction by `pg-migrate`...how do these inner transactions interact with that?

Reply:
You are correct, this is also my understanding! But internal `start transaction`/`commit` queries seem to work too, which comes handy in data-heavy migrations like this one